metrics/metric.go

```diff
@@ -0,0 +1,152 @@
+	Labels            map[string]string
+	LabeledIdentifier string
+	Options           *Options
+	set               *vm.Set
+}
+
+// Options can be used to set advanced metric settings.
+type Options struct {
+	// Name defines an optional human readable name for the metric.
+	Name string
+
+	// AlertLimit defines an upper limit that triggers an alert.
+	AlertLimit float64
+
+	// AlertTimeframe defines an optional timeframe in seconds for which the
+	// AlertLimit should be interpreted in.
+	AlertTimeframe float64
+
+	// Permission defines the permission that is required to read the metric.
+	Permission api.Permission
+
+	// ExpertiseLevel defines the expertise level that the metric is meant for.
+	ExpertiseLevel config.ExpertiseLevel
+
+	// Persist enabled persisting the metric on shutdown and loading the previous
+	// value at start. This is only supported for counters.
+	Persist bool
+}
+
+func newMetricBase(id string, labels map[string]string, opts Options) (*metricBase, error) {
+	// Check formats.
+	if !prometheusFormat.MatchString(id) {
+		return nil, fmt.Errorf("metric name %q must match %s", id, PrometheusFormatRequirement)
+	}
+	for labelName := range labels {
+		if !prometheusFormat.MatchString(labelName) {
+			return nil, fmt.Errorf("metric label name %q must match %s", labelName, PrometheusFormatRequirement)
+		}
+	}
+
+	// Check permission.
+	if opts.Permission < api.PermitAnyone {
+		// Default to PermitUser.
+		opts.Permission = api.PermitUser
+	}
+
+	// Create metric base.
+	base := &metricBase{
+		Identifier: id,
+		Labels:     labels,
+		Options:    &opts,
+		set:        vm.NewSet(),
+	}
+	base.LabeledIdentifier = base.buildLabeledID()
+	return base, nil
+}
+
+// ID returns the given ID of the metric.
+func (m *metricBase) ID() string {
+	return m.Identifier
+}
+
+// LabeledID returns the Prometheus-compatible labeled ID of the metric.
+func (m *metricBase) LabeledID() string {
+	return m.LabeledIdentifier
+}
+
+// Opts returns the metric options. They  may not be modified.
+func (m *metricBase) Opts() *Options {
+	return m.Options
+}
+
+// WritePrometheus writes the metric in the prometheus format to the given writer.
+func (m *metricBase) WritePrometheus(w io.Writer) {
+	m.set.WritePrometheus(w)
+}
+
+func (m *metricBase) buildLabeledID() string {
+	// Because we use the namespace and the global flags here, we need to flag
+	// them as immutable.
+	registryLock.Lock()
+	defer registryLock.Unlock()
+	firstMetricRegistered = true
+
+	// Build ID from Identifier.
+	metricID := strings.TrimSpace(strings.ReplaceAll(m.Identifier, "/", "_"))
+
+	// Add namespace to ID.
+	if metricNamespace != "" {
+		metricID = metricNamespace + "_" + metricID
+	}
+
+	// Return now if no labels are defined.
+	if len(globalLabels) == 0 && len(m.Labels) == 0 {
+		return metricID
+	}
+
+	// Add global labels to the custom ones.
+	// This overrides conflicts.
+	for labelName, labelValue := range globalLabels {
+		m.Labels[labelName] = labelValue
+	}
```

Comment:
Wouldn't it be better if "custom"/local labels have more priority than (and thus overwrite) global labels.?